src/Language/DifferentialDatalog/Parse.hs

```diff
@@ -191,37 +206,49 @@ spec preamble = do
-    <|> (SpRelation     <$> relation)
-    <|> (SpFunc         <$> func)
-    <|> (SpRule         <$> rule)
-    <|> (SpStatement    <$> parseForStatement)
-
-typeDef = withPos $ (TypeDef nopos) <$ reserved "typedef" <*> identifier <*>
-                                       (option [] (symbol "<" *> (commaSep $ symbol "'" *> typevarIdent) <* symbol ">")) <*>
-                                       (optionMaybe $ reservedOp "=" *> typeSpec)
-
-func = withPos $ Function nopos <$  reserved "function"
-                                <*> funcIdent
-                                <*> (parens $ commaSep arg)
-                                <*> (colon *> typeSpecSimple)
-                                <*> (optionMaybe $ reservedOp "=" *> expr)
-
-relation = withPos $ Relation nopos <$> ((True <$ reserved "ground" <* reserved "relation")
-                                         <|>
-                                         (False <$ reserved "relation"))
-                                       <*> relIdent <*> (parens $ commaSep arg)
-
+decl = withPosMany $
+        (return . SpType)      <$> typeDef
+    <|> relation
+    <|> (return . SpFunc)      <$> func
+    <|> (return . SpRule)      <$> rule
+    <|> (return . SpStatement) <$> parseForStatement
+
+typeDef = (TypeDef nopos) <$ reserved "typedef" <*> identifier <*>
+                             (option [] (symbol "<" *> (commaSep $ symbol "'" *> typevarIdent) <* symbol ">")) <*>
+                             (optionMaybe $ reservedOp "=" *> typeSpec)
+
+func = Function nopos <$  reserved "function"
+                      <*> funcIdent
+                      <*> (parens $ commaSep arg)
+                      <*> (colon *> typeSpecSimple)
+                      <*> (optionMaybe $ reservedOp "=" *> expr)
+
+relation = do
+    ground <-  True <$ reserved "ground" <* reserved "relation" 
+           <|> False <$ reserved "relation"
+    relName <- relIdent
+    ((do start <- getPosition
+         fields <- parens $ commaSep arg
+         end <- getPosition
+         let p = (start, end)
+         let tspec = TStruct p [Constructor p relName fields]
+         let tdef = TypeDef nopos relName [] $ Just tspec
+         let rel = Relation nopos ground relName $ TUser p relName []
+         return [SpType tdef, SpRelation rel])
+      <|>
+       (do rel <- brackets $ Relation nopos ground relName <$> typeSpecSimple
+           return [SpRelation rel]))
+        
 arg = withPos $ (Field nopos) <$> varIdent <*> (colon *> typeSpecSimple)
 
-parseForStatement = withPos $ ForStatement nopos <$ reserved "for"
-                                                 <*> (symbol "(" *> expr)
-                                                 <*> (reserved "in" *> relIdent)
-                                                 <*> (optionMaybe (reserved "if" *> expr))
-                                                 <*> (symbol ")" *> statement)
+parseForStatement = ForStatement nopos <$ reserved "for"
```

Comment:
Why remove the withPos?

Reply:
Or if you want to remove it here why not elsewhere as well?

Reply:
I implemented `WithPos` for `SpecItem`, so now I only need to invoke `withPos` once in `decl` and not for each type of `SpecItem`